projects/graph/src/App.js

```diff
@@ -28,25 +62,118 @@ class GraphView extends Component {
     let ctx = canvas.getContext('2d');
     
     // Clear it
     ctx.fillStyle = 'white';
     ctx.fillRect(0, 0, canvasWidth, canvasHeight);
+    ctx.font="20px Georgia";
+    ctx.textAlign = "center";
+    
+    
 
-    // !!! IMPLEMENT ME
-    // compute connected components
-    // draw edges
-    // draw verts
-    // draw vert values (labels)
+    const connectedGraph = graph.getConnectedComponents()
+    ctx.lineWidth = 2;
+    for (let i = 0; i < connectedGraph.length; i++) {
+      const getRandomColor = () => {
+        var letters = '0123456789ABCDEF';
+        var color = '#';
+        for (var i = 0; i < 6; i++) {
+          color += letters[Math.floor(Math.random() * 16)];
+        }
+        return color;
+      }
+      let color = getRandomColor()
+      let color2 = invertColor(color)
+      for (let x = 0; x < connectedGraph[i].length; x++) {
+        ctx.strokeStyle = color
+        let vertex = connectedGraph[i][x];
+        for (let y = 0; y < vertex.edges.length; y++) {
+          const findPos = (startX, startY, endX, endY, distance = 0) => {
+            const deltaX = endX - startX;
+            const deltaY = endY - startY;
+            if (distance === 0) {
+              distance = Math.sqrt((deltaX * deltaX) + (deltaY + deltaY)) / 2
+            }
+            const rad = Math.atan2(deltaY, deltaX);
+            const posX = (distance) * Math.cos(rad) + startX
+            const posY = (distance) * Math.sin(rad) + startY
+            return [posX, posY]
+          }
+          
+          let edge = vertex.edges[y].destination;
+          let start = findPos(vertex.pos.x, vertex.pos.y, edge.pos.x, edge.pos.y, 20)
+          let end = findPos(edge.pos.x, edge.pos.y, vertex.pos.x, vertex.pos.y, 20)
+          let weightXY = [(vertex.pos.x + edge.pos.x)/2, (vertex.pos.y + edge.pos.y)/2]
+          ctx.beginPath();
+          
+          ctx.moveTo(start[0], start[1]);
+          ctx.lineTo(end[0], end[1]);
+          ctx.stroke();
+
+          ctx.beginPath();
+          ctx.fillStyle = color
+          ctx.ellipse(weightXY[0], weightXY[1], 8, 8,  2*Math.PI, 2*Math.PI,  0,  2*Math.PI)
+          ctx.fill()
+          ctx.stroke()
+          ctx.font="12px Monospace";
+          ctx.fillStyle = color2
+          ctx.fillText(vertex.edges[y].weight, weightXY[0], weightXY[1] + 4);
+          ctx.fillStyle = color
+        }
+        ctx.beginPath()
+        
+        ctx.arc(vertex.pos.x, vertex.pos.y, 20, 0, 2*Math.PI)
+        ctx.fillStyle = color2
+        ctx.fill()
+        ctx.stroke()
+        ctx.fillStyle = color
+        ctx.strokeStyle = color
+        ctx.font="20px Georgia";
+        ctx.fillText(vertex.value, vertex.pos.x, vertex.pos.y + 6);
+        ctx.strokeText(vertex.value, vertex.pos.x, vertex.pos.y + 6);
+      }
+    }
   }
   
+  handleButton() {
+    let newGraph = new Graph()
+    newGraph.randomize(5, 4, 150, 0.6)
+    this.setState({
+      graph: newGraph
+    })
+  }
+
+  handleMouse(e) {
```

Comment:
Although this function is incomplete, I was able to find the vertex that is highlighted.